pkg/gpu/containers/containers.go

```diff
@@ -13,22 +13,99 @@ import (
 	"fmt"
 	"regexp"
 	"strconv"
+	"strings"
 
 	workloadmeta "github.com/DataDog/datadog-agent/comp/core/workloadmeta/def"
 	ddnvml "github.com/DataDog/datadog-agent/pkg/gpu/safenvml"
 	gpuutil "github.com/DataDog/datadog-agent/pkg/util/gpu"
+	"github.com/DataDog/datadog-agent/pkg/util/kernel"
 )
 
 // ErrCannotMatchDevice is returned when a device cannot be matched to a container
 var ErrCannotMatchDevice = errors.New("cannot find matching device")
 var numberedResourceRegex = regexp.MustCompile(`^nvidia([0-9]+)$`)
 
+const (
+	nvidiaVisibleDevicesEnvVar = "NVIDIA_VISIBLE_DEVICES"
+)
+
+// HasGPUs returns true if the container has GPUs assigned to it.
+func HasGPUs(container *workloadmeta.Container) bool {
+	switch container.Runtime {
+	case workloadmeta.ContainerRuntimeDocker:
+		// If we have an error, we assume there are no GPUs for the container, so
+		// ignore it.
+		envVar, _ := getDockerVisibleDevicesEnv(container)
+		return envVar != ""
+	default:
+		// We have no specific support for other runtimes, so fall back to the Kubernetes device
+		// assignment if it's there
+		for _, resource := range container.ResolvedAllocatedResources {
+			if gpuutil.IsNvidiaKubernetesResource(resource.Name) {
+				return true
+			}
+		}
+		return false
+	}
+}
+
 // MatchContainerDevices matches the devices assigned to a container to the list of available devices
 // It returns a list of devices that are assigned to the container, and an error if any of the devices cannot be matched
 func MatchContainerDevices(container *workloadmeta.Container, devices []ddnvml.Device) ([]ddnvml.Device, error) {
+	switch container.Runtime {
+	case workloadmeta.ContainerRuntimeDocker:
+		return matchDockerDevices(container, devices)
+	default:
+		// We have no specific support for other runtimes, so fall back to the Kubernetes device
+		// assignment if it's there
+		return matchKubernetesDevices(container, devices)
+	}
+}
+
+func getDockerVisibleDevicesEnv(container *workloadmeta.Container) (string, error) {
+	// We can't use container.EnvVars as it doesn't contain the environment variables
+	// added by the container runtime. We need to get them from the main PID environment.
+	envVar, err := kernel.GetProcessEnvVariable(container.PID, kernel.ProcFSRoot(), nvidiaVisibleDevicesEnvVar)
+	if err != nil {
+		return "", fmt.Errorf("error getting %s for container %s: %w", nvidiaVisibleDevicesEnvVar, container.ID, err)
+	}
+	return strings.TrimSpace(envVar), nil
+}
+
+func matchDockerDevices(container *workloadmeta.Container, devices []ddnvml.Device) ([]ddnvml.Device, error) {
 	var filteredDevices []ddnvml.Device
+	var multiErr error
+
+	visibleDevicesVar, err := getDockerVisibleDevicesEnv(container)
+	if err != nil {
+		return nil, err
+	}
+
+	if visibleDevicesVar == "" {
+		return nil, fmt.Errorf("%s is not set, can't match devices", nvidiaVisibleDevicesEnvVar)
+	}
```

Comment:
do you plan in the future to add support matching the devices by parsing the specs using container runtime API?